velox/functions/prestosql/Repeat.cpp

```diff
@@ -0,0 +1,146 @@
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+#include "velox/expression/DecodedArgs.h"
+#include "velox/expression/VectorFunction.h"
+
+namespace facebook::velox::functions {
+namespace {
+
+// See documentation at https://prestodb.io/docs/current/functions/array.html
+class RepeatFunction : public exec::VectorFunction {
+ public:
+  static constexpr int32_t kMaxResultEntries = 10'000;
+
+  bool isDefaultNullBehavior() const override {
+    // repeat(null, n) returns an array of n nulls.
+    return false;
+  }
+
+  void apply(
+      const SelectivityVector& rows,
+      std::vector<VectorPtr>& args,
+      const TypePtr& outputType,
+      exec::EvalCtx& context,
+      VectorPtr& result) const override {
+    const auto numRows = rows.size();
+    auto constantCount = 0;
+    auto totalCount = 0;
+    exec::DecodedArgs decodedArgs(rows, args, context);
+    DecodedVector* countDecoded = nullptr;
+
+    auto pool = context.pool();
+    BufferPtr nulls;
+    uint64_t* rawNulls = nullptr;
+    if (args[1]->isConstantEncoding()) {
+      if (args[1]->as<ConstantVector<int32_t>>()->isNullAt(0)) {
+        // If count is a null constant, the result should be all nulls.
+        context.moveOrCopyResult(
+            BaseVector::createNullConstant(outputType, numRows, pool),
+            rows,
+            result);
+        return;
+      }
+      constantCount = args[1]->as<ConstantVector<int32_t>>()->valueAt(0);
+      checkCount(constantCount);
+      totalCount = constantCount * numRows;
+    } else {
+      countDecoded = decodedArgs.at(1);
+      context.applyToSelectedNoThrow(rows, [&](auto row) {
+        auto count = countDecoded->isNullAt(row)
+            ? 0
+            : countDecoded->valueAt<int32_t>(row);
+        checkCount(count);
+        totalCount += count;
+      });
+      // Allocate new vector for nulls if necessary.
+      if (countDecoded->mayHaveNulls()) {
+        nulls = allocateNulls(numRows, pool);
+        rawNulls = nulls->asMutable<uint64_t>();
+      }
+    }
+
+    // Allocate new vectors for indices, lengths and offsets.
+    BufferPtr indices = allocateIndices(totalCount, pool);
+    BufferPtr sizes = allocateSizes(numRows, pool);
+    BufferPtr offsets = allocateOffsets(numRows, pool);
+    auto rawIndices = indices->asMutable<vector_size_t>();
+    auto rawSizes = sizes->asMutable<vector_size_t>();
+    auto rawOffsets = offsets->asMutable<vector_size_t>();
+
+    vector_size_t offset = 0;
+    context.applyToSelectedNoThrow(rows, [&](auto row) {
+      if (rawNulls && countDecoded->isNullAt(row)) {
+        // Returns null if count argument is null.
+        // e.g. repeat('a', null) -> null
+        bits::setNull(rawNulls, row);
+        return;
+      }
+
+      int count;
+      if (args[1]->isConstantEncoding()) {
```

Comment:
Constant count is a very common use cases. Currently, this case incurs a lot of unnecessary per-row costs. Let's write a separate loop for that case.